[Step2] 날씨의 아이

July911 opened the pull request:
안녕하세요 릴리 ! :) 

선뜻 리뷰해주신다고해서 감사합니다 하하 
이번 Step2에서는 UseCase를 테스트해보려구합니다
UseCase는 앱의 주요한 기능들이 들어가고 Domain Layer에 위치해있어서 다른 layer와 상관없이 테스트 할 수 있어서 좋다라고 공부해서
UseCase를 테스트해보려구 합니다.


## 🐥 구현한 부분 

![스크린샷 2022-04-18 오후 12 03 16](https://user-images.githubusercontent.com/90888402/163747870-486a015d-bd4f-4c27-a1ed-9def50762e6d.jpg)

각 UseCase별로 테스트 파일을 나누고 

![스크린샷 2022-04-18 오후 12 03 51](https://user-images.githubusercontent.com/90888402/163747930-fafc0b65-09c9-41f7-aa7c-50d15db67dea.jpg)

이런식으로 `sut`를 사용하여 테스트에 용이하게 만들었습니다 :) 

## 🐥 궁금한점 

### 1️⃣  ViewModel 테스트 

UseCase말고 ViewModel 단에서 화면(VC)로 보내 줄 데이터로 가공시키는 작업이 있는데, 이러한 부분은 어떻게 테스트를 해야할지 애매한 부분이 있더라구요 이 부분에 대하여 어떻게 생각하시는지, 릴리라면 어떻게 테스트를 하실지 궁금합니다. 



### 2️⃣  Coordinator 테스트 

Coordinator 에서 앱 전체적으로 공용으로 쓰이는 
```swift
final class MainCoordinator {

    private let navigationController: UINavigationController
    private let imageCacheUseCase = ImageCacheUseCase(
        imageProvideRepository: DefaultImageProvideRepository()
    )
```

ImageCacheUseCase가 ,  Coordinator와 같은 생명주기를 지니는지 테스트를 해보려고 하였으나 
private이 걸려있고 불러올 방법을 찾지 못했습니다. 이러한 경우에 릴리는 어떤식으로 테스트를 진행하시는지 궁금합니다. 

내용은 짧지만 Test는 중요하다고 생각이 들어서 .. 잘부탁드립니다 ! :)

July911 committed:
test: Test bundle 생성 및 ChildOfWeather import

July911 committed:
test: CitySearchRepository 테스트 및 DTO가 Model로 변환되는지 테스트

- 아무것도 입력받지 않았을 때 데이터를 전부 가져오는지 테스트
- 값을 입력했을때, 정확한 값이 Search되는지 테스트
- 변환을 담당해주는 DTO의 toDomain메서드가 정상적으로 동작하는지 테스트

July911 committed:
test: ImageCacheUseCase 테스트

- ImageCacheData로 묶인 key와 value를 넣어 캐싱
- key값을 입력했을때 제대로 fetch되는지 테스트

July911 committed:
test: AddressSearchUseCase 테스트

- 이매동의 위경도를 넣었을때, 제대로된 주소값이 나오는지 테스트
- 비동기작업의 테스트를 위해 promise,fulfill과 wait을 사용

July911 committed:
test: 각 UseCaes별로 파일 분리

- 각 파일별 SystemUnderTest를 정의하여 테스트에 용이하게 변경

July911 committed:
test: DetailShowUseCase의 fetchURL메서드에 주소를 넣었을때, URL로 제대로 변환되는지 test

yeahg-dev commented on ChildOfWeatherTests/UseCaseTests/DetailShowUseCaseTests.swift:
```diff
@@ -0,0 +1,22 @@
+    var sut: DetailShowUseCase?
+
+    override func setUpWithError() throws {
+        let repository = DefaultWeatherRepository(service: APIService())
```
네트워킹의 상태에 영향을 받고있는 테스트 같습니다.
네트워크와 무관하도록 개선해보는 것도 좋을 것 같아요🙂

July911 replied:
APIService를 `URLSessionNetworkService`프로토콜로 묶고 TestDouble로 MockAPIService를 만들어 
테스트를 진행했습니다 :) 
테스트를 하는데 네트워킹 통신을 하면 안된다고 생각합니다 ! 감사합니다.

yeahg-dev commented on ChildOfWeatherTests/UseCaseTests/CitySearchUseCaseTests.swift:
```diff
@@ -0,0 +1,46 @@
+        XCTAssertEqual(cities, cityfromKr)
+    }
+    
+    func test_CitySearchUseCase_yongin을_search했을때_값은_yongin이다() {
```
존재하지 않는 도시명을 검색했을 경우의 테스트는 진행하지 않은 이유가 있을까요?

July911 replied:
좋은 지적 감사합니다. 성공케이스만이 아니라 실패케이스도 
`XCTAssertNil` 메서드를 통해 검증했습니다 :)

yeahg-dev commented on ChildOfWeatherTests/UseCaseTests/AddressSearchUseCaseTests.swift:
```diff
@@ -0,0 +1,30 @@
+import XCTest
+@testable import ChildOfWeather
+
+class AddressSearchUseCaseTests: XCTestCase {
```
테스트와 무관한 질문입니다. (저스트 큐리어스,,,🧐)
`AdressSearchUseCase`는 다른 유스케이스와 같이 뷰단이 참조하고 있지 않고, 델리게이트도 없다는 점에서 다른 뷰모델과 조금 구조가 다르다고 생각했는데요. 그래서 Util로 빼어도 괜찮을 것 같다고 생각했습니다. 
`AdressSearchUseCase`를 이와 같이 설계하신 이유가 궁금합니다!

July911 replied:
좋은 질문인거같습니다 ;) 감사합니다 
먼저 AddressSearchUseCase는 Repository단에서 `CLGeocoder`에 인스턴스를 가지고있고 이걸 계속 참고하여 사용하고 있습니다. 
릴리가 말씀하신것처럼 Util로 빼는 방법도 성능적인 측면에서 더 좋다라고 생각합니다.
하지만 그럼에도 이렇게 쓴 이유는 ViewModel에서 View에 전달하기위해 필요한 URL을 생성하기 위해서 각각의 UseCase의 기능을 종합해서 사용하고 있는데, ViewModel이 UseCase나 Delegate , Coordinator를 제외한 다른것을 들고 있는것이 어색해보여서 이렇게 하나의 기능으로 뺐던거 같습니다.
릴리 생각은 앱 전역에서 쓰이는 데이터고 주소값을 받아오는 역할밖에 하지 않기때문에 싱글톤이나 extension으로 만들어 쓰라는 말씀이신건가요 ?

yeahg-dev replied:
비지니스로직을 내포하고 있지 않고 있다는 점에서 유스케이스보다는 유틸적인 색채를 띄고 있지 않나 싶어서 말씀드렸습니당!
저라면 싱글톤이나 익스텐션에서 구현하는 방법도 고려해봤을 것 같아요!

yeahg-dev reviewed (commented):
안녕하세요 줄라이~
리뷰 할 기회를 주셔서 감사합니다😊
덕분에 저도 새로운 고민과 공부를 하게 되었어요ㅎㅎ
각 유스케이스에 대한 테스트를 잘 작성해주셨네요!
수고많으셨습니다👍

궁금한 점에 대한 제 의견 공유드립니다

### 1️⃣ ViewModel 테스트
현재 설계 구조상으로 봤을 땐, 저라면 테스트를 위한 mock델리게이트 객체를 생성해서  ViewModel에서 가공한 데이터를 체킹하도록 테스트 해볼 것 같습니다. 

### 2️⃣ Coordinator 테스트
`ImageCacheUseCase`가 `Coordinator`의 deinit시점에 함께 deinit되는지를 테스트 해보고 싶은걸까요? 
제가 질문을 잘 이해한건지 모르겠네요😅 
추가로 코멘트 남겨주시면 같이 고민해보겠습니다!

July911 committed:
test: TestDouble로 MockAPIService를 생성 및 테스트에 적용

July911 committed:
refactor: AddressSearchUseCase 삭제 및 LocationManager 싱글톤 생성

- Repository 단에서 CLGeocoder생성만 담당하여 이를 삭제
- CLGeocoder의 초기화를 매번해줘야하는 비용과 싱글톤의 메모리 유지 비용을 트레이드 오프

yeahg-dev commented:
수고하셨어요!!!!!🚀

다음 스텝도 화이팅하세요옷~!
